### PR TITLE
Boolean logic AND and OR to circuit library

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

The boolean logical operations AND and OR are implemented as functions since they were moved over from Aqua. This PR converts these to circuit library objects.

### Details and comments

The circuit methods `QuantumCircuit.AND` and `OR` are still kept for deprecation.

### CODE REVIEW

File: test/python/circuit/test_library.py [hunk 1]
```diff
@@ -22,15 +22,46 @@
...
+
+        if hasattr(boolean_circuit, 'num_ancilla_qubits'):
+            num_ancillas = boolean_circuit.num_ancilla_qubits
+        else:
+            num_ancillas = 0
+
+        probabilities = defaultdict(float)
+        for i, statevector_amplitude in enumerate(statevector):
+            i = bin(i)[2:].zfill(circuit.num_qubits)[num_ancillas:]
+            probabilities[i] += np.real(np.abs(statevector_amplitude) ** 2)
```
Comment (ajavadia) [inline test/python/circuit/test_library.py:52]: this whole block can be replaced by something like:
```python
from qiskit.quantum_info import Statevector
sv = Statevector.from_label('0' * circuit.num_qubit)
sv = sv.evolve(circuit)
probabilities = sv.probabilities()
```
though i'm missing the num_ancillas part, not quite sure what that does.

File: test/python/circuit/test_library.py [hunk 2]
```diff
@@ -22,15 +22,46 @@
...
 from qiskit.circuit.library.arithmetic import (LinearPauliRotations, PolynomialPauliRotations,
                                                IntegerComparator, PiecewiseLinearPauliRotations,
                                                WeightedAdder)
 
 
+@ddt
 class TestBooleanLogicLibrary(QiskitTestCase):
     """Test library of boolean logic quantum circuits."""
 
+    def assertBooleanFunctionIsCorrect(self, boolean_circuit, reference):
```
Comment (ajavadia) [inline test/python/circuit/test_library.py:35]: this is nice, we should make a note to rewrite the other boolean logic tests using this.

File: test/python/circuit/test_library.py (hunk 1 above)
Comment (Cryoris) [reply test/python/circuit/test_library.py:52]: That's tracing out the ancillas 🙂 but the `Statevector` class can also do this I think using the `qargs` argument

File: test/python/circuit/test_library.py (hunk 2 above)
Comment (Cryoris) [reply test/python/circuit/test_library.py:36]: I added notes to the other tests

**Pull request merged**